src/builtins/host_abi.zig

```diff
@@ -10,57 +10,53 @@
+    /// Roc function to write its entire return value into the address.
+    *anyopaque,
+    /// Varargs, all of which are pointers (to arguments that will be passed
+    /// into the Roc function). Arguments with small sizes can be combined into
+    /// a struct, and a pointer to that struct can be passed as one of these.
+    anytype,
+) callconv(.C) void;
 
 /// The operations (in the form of function pointers) that a running Roc program
 /// needs the host to provide.
 ///
 /// This is used in both calls from actual hosts as well as evaluation of constants
 /// inside the Roc compiler itself.
-pub fn RocOps(comptime CallEnv: type, comptime HostFns: type) type {
-    return struct {
-        /// The host provides this pointer, and Roc passes it to each of the following
-        /// function pointers as a second argument. This lets the host do things like use
-        /// arena allocators for allocation and deallocation (by putting the arena in here).
-        /// The pointer can be to absolutely anything the host likes, or null if unused.
-        env: *CallEnv,
-        /// Similar to  _aligned_malloc - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc
-        roc_alloc: fn (*RocAlloc, *CallEnv) void,
-        /// Similar to  _aligned_free - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-free
-        roc_dealloc: fn (*RocDealloc, *CallEnv) void,
-        /// Similar to  _aligned_realloc - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-realloc
-        roc_realloc: fn (*RocRealloc, *CallEnv) void,
-        /// Called when the Roc program has called `dbg` on something.
-        roc_dbg: fn (*RocDbg, *CallEnv) void,
-        /// Called when the Roc program has run an `expect` which failed.
-        roc_expect_failed: fn (*RocExpectFailed, *CallEnv) void,
-        /// Called when the Roc program crashes, e.g. due to integer overflow.
-        /// This function must not return, because the Roc program assumes it will
-        /// not continue to be executed after this function is called.
-        roc_crashed: fn (*RocCrashed, *CallEnv) void,
-        /// At the end of this struct, the host must include all the functions
-        /// it wants to provide to the Roc program for the Roc program to call
-        /// (e.g. I/O operations and such).
-        host_fns: HostFns,
-    };
-}
+pub const RocOps = struct {
+    /// The host provides this pointer, and Roc passes it to each of the following
+    /// function pointers as a second argument. This lets the host do things like use
+    /// arena allocators for allocation and deallocation (by putting the arena in here).
+    /// The pointer can be to absolutely anything the host likes, or null if unused.
+    env: *anyopaque,
+    /// Similar to  _aligned_malloc - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc
+    roc_alloc: *const fn (*RocAlloc, *anyopaque) callconv(.C) void,
```

Comment:
Of note on ABI as we were discussing some on zulip. In `RocCall` above, there is an explicitly `*anyopaque` for the return value.

Yet for all of the roc builtin functoins, we make the return type part of the struct instead of an `*anyopaque`. I don't think it really matters either way, but that is the oddity I was mentioning.